### ISSUE TITLE
User routine for func_callback returned -500

### ISSUE THREAD

**Issue by ferrolho** [JuliaOpt/KNITRO.jl#120]
After calling `KNITRO.KN_solve(kc)` the optimisation process is interrupted with:
```
ERROR: User routine for func_callback returned -500.
       Could not evaluate objective or constraints at the current point.

EXIT: Callback function error.
```

I must be doing something wrong in one of the callbacks, but I am finding it very hard to understand what I am doing wrong. There is no backtrace or any other printed information that could hint me to what I could possibly be doing wrong.

Is there any way I could go on to debug this one?

**Comment by frapac** [comment]
I have sometimes encountered this issue when using callbacks defined in closures. The workaround I have is to precompile the callbacks before sending them to Knitro, e.g. 
```julia
my_func(x) = ...
my_func(zeros(n_dim)) #precompilation

KN_add_objective_callback(kc, my_func)
```

Do you have a MWE that would allow to investigate this issue?

**Comment by ferrolho** [comment]
It is not so easy for me to give a MWE because my code is quite extensive. It's a bit curious though, because I have not had this issue in the past.

Long story short: I was developing something and was lazy enough to not care much about organisation and was doing everything in a _single_ jupyter notebook; and I never had any error like the one in the OP. Now I am trying to move away from that enormous notebook and put most of the things in a julia module of its own. It was at this point that I started to have this issue.

I think it may have to do with the way I am rebuilding the code structure. In the notebook, most of the things I created were obviously global variables within that notebook. Now that I am trying to refactor the code, the way I am first approaching the task is to have this `mutable struct Problem` holding most of those variables from the "old" notebook. In the beginning of the "new" notebook, a global variable `problem` is created, instantiating the struct. I think this may be the possible point of failure, somehow...

**Comment by ferrolho** [comment]
I have been trying to narrow the problem down but this still is an issue. I have noticed that if I run this the first time, it will just not do anything at all (optimisation-wise); for the second time, it will fail later, maybe after one or two iterations; and for the third time I run the optimisation, it usually is able to complete without any problem... Very weird.

I have also followed your suggestion of forcing precompilation, but it has not worked for me.

**Comment by ferrolho** [comment]
Okay, I am getting closer. I can tell it is failing in [`eval_fc_wrapper()`](https://github.com/JuliaOpt/KNITRO.jl/blob/c84355c10803cd3b10b540ab1d2ab64fa3750f38/src/kn_callbacks.jl#L285-L315); more specifically, here:
https://github.com/JuliaOpt/KNITRO.jl/blob/c84355c10803cd3b10b540ab1d2ab64fa3750f38/src/kn_callbacks.jl#L300-L307

At some point, one of the iterations calls a function/gradient evaluation with my custom callback, which in turn triggers the condition in _line 305_, finally returning `Cint(KN_RC_CALLBACK_ERR)`.

Going one level up, and looking at what is calling `eval_fc_wrapper()`, we find:
https://github.com/JuliaOpt/KNITRO.jl/blob/c84355c10803cd3b10b540ab1d2ab64fa3750f38/src/kn_callbacks.jl#L479-L504

So, there is this macro `@cfunction` which is wrapping `eval_fc_wrapper`. I think the issue is right here.

Here is what the documentation says about  `@cfunction`:
> Generate a C-callable function pointer from the Julia function `closure` for the given type signature. To pass the return value to a `ccall`, use the argument type `Ptr{Cvoid}` in the signature.
> 
> Note that the argument type tuple must be a literal tuple, and not a tuple-valued variable or expression (although it can include a splat expression). And that these arguments will be evaluated in global scope during compile-time (not deferred until runtime). Adding a '$' in front of the function argument changes this to instead create a runtime closure over the local variable `callable`.
> 
> Source: https://docs.julialang.org/en/v1/base/c/#Base.@cfunction

_"(...) these arguments will be evaluated in global scope during compile-time (not deferred until runtime)."_

I am not sure if this is the reason why it is failing... Any help would be greatly appreciated!

**Comment by frapac** [comment]
Interesting. I really appreciate your detailed investigation, I think that will help a lot to debug what's going on. 
My best guess is that the issue is related to the garbage collector (that would explain the non-deterministic behavior observed in your issue). For instance, looking at your stacktrace, we get
```julia
unknown function (ip: 0xffffffffffffffff)
```
The pointer to the Julia's callback becomes `NULL`, thus leading to a crash in Knitro.jl. A solution would be to use the `GC.preserve` macro:
https://docs.julialang.org/en/v1/base/base/#Base.GC.@preserve

I recognize that this issue is related to Knitro.jl. Trouble with the garbage collector were already observed before, e.g. https://github.com/JuliaOpt/Cbc.jl/issues/100

Unfortunately, I do not have access to Knitro this week-end. I will work on a fix on Monday.

**Comment by ferrolho** [comment]
> A solution would be to use the GC.preserve macro.

Yes, thank you! You are correct! I followed your advice and I am now wrapping the call to `KN_solve(kc)` with `GC.@preserve` and for all the `cb` I create with `KN_add_eval_callback()` or similar:

```julia
# The following is a workaround for a bug
# concerning the unwanted GC behaviour.
# Cf. https://github.com/JuliaOpt/KNITRO.jl/issues/120
if true
    GC.@preserve cb_ee cb_dyn begin
        KNITRO.KN_solve(kc)
    end
else
    KNITRO.KN_solve(kc)
end
```

And it works! No crash so far on optimisations with hundreds of iterations and function/gradient calls.

I would like to understand the root cause of this issue and fix it, but for now this will help keeping me go. Thank you very much!

**Comment by frapac** [comment]
Happy to see that workaround works on your example! I will push a fix in `Knitro.jl` next week.

**Comment by ferrolho** [comment]
Great! Looking forward to see the root of the issue and the fix. :smiley: Again: thank you for the help!

**Comment by frapac** [comment]
Closed by #122
